### PR TITLE
refactor and improve server authentication

### PR DESCRIPTION
**Pull request by hannesm**
see individual commit messages for details, this also adds a mode allowing unauthenticated zone transfers in the primary server (which @dinosaure had interest in).

### CODE REVIEW

**Comment by hannesm** [comment]
ready for review

File: server/dns_server.ml [hunk 1]
```diff
@@ -95,17 +91,28 @@ module Authentication = struct
...
+        (fun op -> Domain_name.equal_label lbl (operation_to_string op))
+        all_ops
     in
     match Domain_name.find_label ~rev:true name is_op with
     | None -> Domain_name.host_exn name
     | Some idx ->
       let amount = succ idx in
       Domain_name.host_exn (Domain_name.drop_label_exn ~amount name)
 
+  let key_operation name =
```
[the remark this answers is not in the document]
Comment (hannesm) [reply server/dns_server.ml]: not sure I follow, let me run utop for you:

```OCaml
# rm server/dns_server.mli
# dune utop
utop # let name = Domain_name.of_string_exn "foo.bar._transfer.nevermind._notify.google.com";;
val name : [ `raw ] Domain_name.t = <abstr>
utop # Domain_name.to_string @@ Dns_server.Authentication.key_zone name;;
- : string = "google.com"
Dns_server.Authentication.key_operation name;;
- : [ `Notify | `Transfer | `Update ] option = Some `Notify
```

Returning both from a single function is a good idea, I may try this (figuring out before whether the callers are fine with it).

(NB in a previous commit https://github.com/mirage/ocaml-dns/commit/a270f3f4372250235308870bee60b5f5e0cfdb4a your statement was correct, and key_operation and key_zone were using different search strategies (one from the last label, one from the first label).

[the remark this answers is not in the document]
Comment (hannesm) [reply server/dns_server.ml]: oh, and the other way around (`foo.bar._notify.bar._transfer.google.com`) actually leads to different interpretations (though not harmful due to the order in `all_ops` -- anyways, not very intuitive, working on a fix).

File: server/dns_server.ml [hunk 2]
```diff
@@ -26,30 +26,26 @@ let guard p err = if p then Ok () else Error err
...
+    | `Notify
   ]
+  let all_ops = [ `Notify ; `Transfer ; `Update ]
 
-  type a = Dns_trie.t -> proto -> ?key:[ `raw ] Domain_name.t -> operation ->
-    zone:[ `raw ] Domain_name.t -> bool
+  let access_granted ~required key = match required, key with
+    | _, `Update -> true | `Update, _ -> false
+    | _, `Transfer  -> true | `Transfer, _ -> false
+    | `Notify, `Notify -> true
```
[the remark this answers is not in the document]
Comment (hannesm) [reply server/dns_server.ml]: done in 3afab396f0ba2a5d2cdd30724de057453c923022

File: server/dns_server.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (hannesm) [reply server/dns_server.ml]: see 0958086884dc6c3faa86fe70c33448041011298b which unifies both functions into a single one.

**Comment by dinosaure** [comment]
Thanks!

**Pull request merged**